src/player_activity.cpp

```diff
@@ -318,6 +318,11 @@ void player_activity::finish( player *p )
             p->do_read(&(p->i_at(position)));
             if (type == ACT_NULL) {
                 add_msg(_("You finish reading."));
+
+                // below code added in to support reading from ground/vehicles
+                if ( values[1] != FRM_INV ) {
+                    g->drop(p->inv.size()-1);
```

Comment:
You don't need to do that because the book's address is saved into player_activity's "position" variable already.  I think it's enough to simply replace `g->drop(p->inv.size()-1);` with `g->drop(position);`